File: Camfix_App/lib/models/service_quote.dart
/// One version of a technician's itemized repair quote for a booking, from
/// `GET /api/bookings/{id}/quotes`. A booking can have several — each new
/// version supersedes the previous PENDING one (which becomes `REVISED`).
class ServiceQuote {
  const ServiceQuote({
    required this.id,
    required this.jobId,
    required this.technicianId,
    required this.version,
    required this.inspectionFee,
    required this.laborCost,
    required this.partsCost,
    required this.travelFee,
    required this.totalAmount,
    this.reason,
    required this.status,
    this.createdAt,
  });

  final int id;
  final int jobId;
  final int technicianId;
  final int version;
  final double inspectionFee;
  final double laborCost;
  final double partsCost;
  final double travelFee;
  final double totalAmount;
  final String? reason;
  final String status; // PENDING | ACCEPTED | REJECTED | REVISED | EXPIRED
  final DateTime? createdAt;

  bool get isPending => status == 'PENDING';
  bool get isAccepted => status == 'ACCEPTED';
  bool get isRejected => status == 'REJECTED';

  factory ServiceQuote.fromJson(Map<String, dynamic> j) => ServiceQuote(
        id: (j['id'] as num?)?.toInt() ?? 0,
        jobId: (j['jobId'] as num?)?.toInt() ?? 0,
        technicianId: (j['technicianId'] as num?)?.toInt() ?? 0,
        version: (j['version'] as num?)?.toInt() ?? 1,
        inspectionFee: (j['inspectionFee'] as num?)?.toDouble() ?? 0,
        laborCost: (j['laborCost'] as num?)?.toDouble() ?? 0,
        partsCost: (j['partsCost'] as num?)?.toDouble() ?? 0,
        travelFee: (j['travelFee'] as num?)?.toDouble() ?? 0,
        totalAmount: (j['totalAmount'] as num?)?.toDouble() ?? 0,
        reason: j['reason']?.toString(),
        status: (j['status'] ?? 'PENDING').toString(),
        createdAt: j['createdAt'] == null
            ? null
            : DateTime.tryParse(j['createdAt'].toString())?.toLocal(),
      );
}

File: Backend/src/main/java/com/api/entity/ServiceQuote.java
package com.api.entity;

import java.time.LocalDateTime;

import org.hibernate.annotations.CreationTimestamp;
import org.hibernate.annotations.UpdateTimestamp;

import jakarta.persistence.Column;
import jakarta.persistence.Entity;
import jakarta.persistence.GeneratedValue;
import jakarta.persistence.GenerationType;
import jakarta.persistence.Id;
import jakarta.persistence.Table;
import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Data;
import lombok.NoArgsConstructor;

/**
 * A technician's itemized repair quote for a {@link Job}, versioned so a
 * revised quote is a new row rather than an edit - see {@code Quote_Status}.
 * Kept flat (jobId/technicianId only) to match {@code Job}'s no-JPA-relations style.
 */
@Entity
@Table(name = "service_quote")
@Data
@NoArgsConstructor
@AllArgsConstructor
@Builder
public class ServiceQuote {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @Column(name = "job_id", nullable = false)
    private Long jobId;

    @Column(name = "technician_id", nullable = false)
    private Long technicianId;

    @Column(nullable = false)
    private Integer version;

    @Column(name = "inspection_fee", nullable = false)
    @Builder.Default
    private Double inspectionFee = 0.0;

    @Column(name = "labor_cost", nullable = false)
    @Builder.Default
    private Double laborCost = 0.0;

    @Column(name = "parts_cost", nullable = false)
    @Builder.Default
    private Double partsCost = 0.0;

    @Column(name = "travel_fee", nullable = false)
    @Builder.Default
    private Double travelFee = 0.0;

    /** Always recomputed by the service layer from the four fees above - never
     *  accepted directly from a client request. */
    @Column(name = "total_amount", nullable = false)
    @Builder.Default
    private Double totalAmount = 0.0;

    @Column(length = 500)
    private String reason;

    /** PENDING | ACCEPTED | REJECTED | REVISED | EXPIRED */
    @Column(nullable = false, length = 20)
    private String status;

    @CreationTimestamp
    @Column(name = "created_at", updatable = false)
    private LocalDateTime createdAt;

    @UpdateTimestamp
    @Column(name = "updated_at")
    private LocalDateTime updatedAt;
}
